pinot-plugins/pinot-input-format/pinot-csv/src/main/java/org/apache/pinot/plugin/inputformat/csv/CSVRecordReader.java

```diff
@@ -202,158 +209,142 @@ public Map<String, Integer> getCSVHeaderMap() {
-    try {
-      _nextRecord = null;
-      GenericRow genericRow = new GenericRow();
-      readNextLine(genericRow);
-      _nextRecord = genericRow;
-    } catch (Exception e) {
-      LOGGER.info("Error parsing next record.", e);
+  private void closeIterator()
+      throws IOException {
+    // if header is not provided by the client it would be rebuilt. When it's provided by the client it's initialized
+    // once in the constructor
+    if (useLineIterator(_config) && !_isHeaderProvided) {
+      _headerMap.clear();
+    }
+
+    if (_bufferedReader != null) {
+      _bufferedReader.close();
     }
-    return _nextRecord != null;
   }
 
-  private void readNextLine(GenericRow reuse)
-      throws IOException {
-    while (_nextLine != null) {
-      try (Reader reader = new StringReader(_nextLine)) {
-        try (CSVParser csvParser = _format.parse(reader)) {
-          List<CSVRecord> csvRecords = csvParser.getRecords();
-          if (csvRecords != null && csvRecords.size() > 0) {
-            // There would be only one record as lines are read one after the other
-            CSVRecord record = csvRecords.get(0);
-            _recordExtractor.extract(record, reuse);
-            break;
-          } else {
-            // Can be thrown on: 1) Empty lines 2) Commented lines
-            throw new NoSuchElementException("Failed to find any records");
+  class LineIterator implements Iterator<CSVRecord> {
+    private final boolean _skipHeaderRecord;
+
+    private String _nextLine;
+
+    private CSVRecord _current;
+
+    public LineIterator(CSVRecordReaderConfig config) {
+      _skipHeaderRecord = config.isSkipHeader();
+
+      init();
+    }
+
+    private void init() {
+      try {
+        if (_isHeaderProvided) {
+          if (_skipHeaderRecord) {
+            // When skip header config is set and header is supplied – skip the first line from the input file
+            _bufferedReader.readLine();
+            // turn off the property so that it doesn't interfere with further parsing
+            _format = _format.builder().setSkipHeaderRecord(false).build();
           }
-        } catch (Exception e) {
-          _skippedLinesCount++;
-          LOGGER.debug("Skipped input line: {} from file: {}", _nextLine, _dataFile, e);
-          // Find the next line that can be parsed
-          _nextLine = _bufferedReader.readLine();
+        } else {
+          // read the first line
+          String headerLine = _bufferedReader.readLine();
+          _headerMap = parseLineAsHeader(_format, headerLine);
+          _format = _format.builder()
+              // If header isn't provided, the first line would be set as header and the 'skipHeader' property
+              // is set to false.
+              .setSkipHeaderRecord(false).setHeader(_headerMap.keySet().toArray(new String[0])).build();
         }
+        _nextLine = _bufferedReader.readLine();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
```

Comment:
There's no clean up needed on failure (eg: for _bufferedReader)

Reply:
yes. This is existing logic.

I assume, that `RecordReader` being `Closeable` is closed post an error? Is that not the case, then it is a bigger issue. The `close()` method closes the used resources

Reply:
Yeah caller will close the RecordReader i think